test/test.c

```diff
@@ -28636,6 +28639,495 @@ void test_JetStreamSubscribePull(void)
+        args.closed = false;
+        args.sum = 0;
+        jsSubOptions_Init(&so);
+        jsOptions_Init(&jsOpts);
+        so.Config.MaxAckPending = 10;
+        so.Config.AckWait = NATS_MILLIS_TO_NANOS(300);
+        so.Config.MaxRequestMaxBytes = 777;
+        so.ManualAck = false;
+        jsOpts.PullSubscribeAsync.CompleteHandler = _completePullAsync;
+        jsOpts.PullSubscribeAsync.CompleteHandlerClosure = &args;
+        jsOpts.PullSubscribeAsync.FetchSize = tc->fetchSize;
+        jsFetchRequest lifetime = {
+            .Batch = tc->want,
+            .MaxBytes = tc->maxBytes,
+            .NoWait = tc->noWait,
+            .Expires = NATS_MILLIS_TO_NANOS(tc->expires),
+        };
+        s = js_PullSubscribeAsync(&sub, js, "foo", "dur", _recvPullAsync, &args, &lifetime, &jsOpts, &so, &jerr);
+        if ((s != NATS_OK) && (sub == NULL) && (jerr != 0))
+            FAIL("Failed to create pull subscription, unusual");
+        natsMutex_Unlock(args.m);
+
+        for (int i = 0; (s == NATS_OK) && (i < tc->during); i++)
+            s = js_Publish(NULL, js, "foo", "hello", 5, NULL, &jerr);
+        if (s != NATS_OK)
+            FAIL("Failed to publish, unusual");
+
+        testf("%s: ", tc->name);
+        testCond(_testBatchCompleted(
+            &args, sub, batchWaitTimeout, tc->expectedStatus, tc->expectedN, tc->orFewer));
+
+        natsSubscription_Destroy(sub);
+        sub = NULL;
+    }
+
+    JS_TEARDOWN;
+    _destroyDefaultThreadArgs(&args);
+}
+
+void test_JetStreamSubscribePullAsyncMissedHB(void)
+{
+    natsStatus s;
+    natsSubscription *sub = NULL;
+    jsErrCode jerr = 0;
+    jsStreamConfig sc;
+    jsOptions jsOpts;
+    jsSubOptions so;
+    struct threadArg args;
+
+    JS_SETUP(2, 9, 2);
+
+    s = _createDefaultThreadArgsForCbTests(&args);
+    if (s != NATS_OK)
+        FAIL("Unable to setup test");
+
+    test("Create stream: ");
+    jsStreamConfig_Init(&sc);
+    sc.Name = "TEST";
+    sc.Subjects = (const char *[1]){"foo"};
+    sc.SubjectsLen = 1;
+    s = js_AddStream(NULL, js, &sc, NULL, &jerr);
+    testCond((s == NATS_OK) && (jerr == 0));
+
+    test("Check invalid heartbeat : ");
+    natsMutex_Lock(args.m);
+    args.control = 0;      // don't ack, will be auto-ack
+    args.status = NATS_OK; // batch exit status will be here
+    args.msgReceived = false;
+    args.closed = false;
+    args.sum = 0;
+    jsSubOptions_Init(&so);
+    jsOptions_Init(&jsOpts);
+    jsOpts.PullSubscribeAsync.CompleteHandler = _completePullAsync;
+    jsOpts.PullSubscribeAsync.CompleteHandlerClosure = &args;
+    jsFetchRequest lifetime = {
+        .Batch = 100,
+        .Expires = NATS_MILLIS_TO_NANOS(1),
+        .Heartbeat = NATS_MILLIS_TO_NANOS(10),
+    };
+    natsMutex_Unlock(args.m);
+
+    s = js_PullSubscribeAsync(&sub, js, "foo", "dur", _recvPullAsync, &args, &lifetime, &jsOpts, &so, &jerr);
+    testCond((s == NATS_OK) && _testBatchCompleted(&args, sub, 100, NATS_ERR, 0, false));
```

Comment:
Again, tests often with s==NATS_OK, but function returns 0.

Reply:
Sorry, did not understand this last one, Windows-specific?

Reply:
What I meant is that `s` is `NATS_OK`, what returns `0` is `_testBatchCompleted`.

Reply:
Resolved in favor of thread https://github.com/nats-io/nats.c/pull/785#discussion_r1725839401